pytennis/tennis.py

```diff
@@ -0,0 +1,489 @@
+from keras.layers import Dense
+from collections import deque
+
+pygame.init()
+
+
+class tennis:
+    def __init__(self, fps=50):
+        self.GeneralReward = False
+        self.net = Network(150, 450, 150, 650)
+        self.updateRewardA = 0
+        self.updateRewardB = 0
+        self.updateIter = 0
+        self.lossA = 0
+        self.lossB = 0
+        self.restart = False
+
+        self.AgentA = DQN()
+        self.AgentB = DQN()
+
+        # Testing
+        self.net = Network(150, 450, 150, 650)
+        self.NetworkA = self.net.network(
+            300, ysource=80, Ynew=650)  # Network A
+        self.NetworkB = self.net.network(
+            200, ysource=650, Ynew=80)  # Network B
+
+        pygame.init()
+        self.BLACK = (0, 0, 0)
+
+        self.myFontA = pygame.font.SysFont("Times New Roman", 25)
+        self.myFontB = pygame.font.SysFont("Times New Roman", 25)
+        self.myFontIter = pygame.font.SysFont('Times New Roman', 25)
+
+        self.FPS = fps
+        self.fpsClock = pygame.time.Clock()
+
+    def setWindow(self):
+
+        # set up the window
+        self.DISPLAYSURF = pygame.display.set_mode((600, 750), 0, 32)
+        pygame.display.set_caption(
+            'REINFORCEMENT LEARNING (DQN) - TABLE TENNIS')
+        # set up the colors
+        self.BLACK = (0, 0, 0)
+        self.WHITE = (255, 255, 255)
+        self.RED = (255, 0, 0)
+        self.GREEN = (0, 255, 0)
+        self.BLUE = (0, 0, 255)
+
+        return
+
+    def display(self):
+        self.setWindow()
+        self.DISPLAYSURF.fill(self.WHITE)
+        pygame.draw.rect(self.DISPLAYSURF, self.BLACK, (50, 100, 500, 550))
+        pygame.draw.rect(self.DISPLAYSURF, self.RED, (50, 365, 500, 20))
+        # pygame.draw.rect(self.DISPLAYSURF, self.BLACK, (0, 20, 600, 20))
+        # pygame.draw.rect(self.DISPLAYSURF, self.BLACK, (0, 770, 600, 20))
+        return
+
+    def reset(self):
+        return
+
+    def evaluate_state_from_last_coordinate(self, c):
+        """
+        cmax: 450
+        cmin: 150
+
+        c definately will be between 150 and 450.
+        state0 - (50 - 99)
+        state1 - (100 - 149)
+        state2 - (150 - 199)
+        state3 - (200 - 249)
+        state4 - (250 - 299)
+        state5 - (300 - 349)
+        state6 - (350 - 399)
+        state7 - (400 - 449)
+        state8 - (450- 499)
+        state9 - (500 - 550)
+        """
+        if c >= 50 and c <= 99:
+            return 0
+        elif c >= 100 and c <= 149:
+            return 1
+        elif c >= 150 and c <= 199:
+            return 2
+        elif c >= 200 and c <= 249:
+            return 3
+        elif c >= 250 and c <= 299:
+            return 4
+        elif c >= 300 and c <= 349:
+            return 5
+        elif c >= 350 and c <= 399:
+            return 6
+        elif c >= 400 and c <= 449:
+            return 7
+        elif c >= 450 and c <= 499:
+            return 8
+        elif c >= 500 and c <= 550:
+            return 9
```

Comment:
We should not calculate the state vector like this. State vector would be the x/y coordinates of ball and bats.